File: FWFramework/Classes/Objc/Service/FWNotification.m
/**
 @header     FWNotification.m
 @indexgroup FWFramework
      FWNotification
 @author     wuyong
 @copyright  Copyright © 2019 wuyong.site. All rights reserved.
 @updated    2019/5/17
 */

#import "FWNotification.h"

@interface FWNotificationManager () <UNUserNotificationCenterDelegate>

@end

@implementation FWNotificationManager

+ (FWNotificationManager *)sharedInstance
{
    static FWNotificationManager *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[FWNotificationManager alloc] init];
    });
    return instance;
}

#pragma mark - Authorize

- (void)authorizeStatus:(void (^)(FWAuthorizeStatus))completion
{
    [[FWAuthorizeManager managerWithType:FWAuthorizeTypeNotifications] authorizeStatus:completion];
}

- (void)requestAuthorize:(void (^)(FWAuthorizeStatus))completion
{
    [[FWAuthorizeManager managerWithType:FWAuthorizeTypeNotifications] authorize:completion];
}

#pragma mark - Badge

- (void)clearNotificationBadges
{
    [UIApplication sharedApplication].applicationIconBadgeNumber = 0;
}

#pragma mark - Handler

- (void)registerNotificationHandler
{
    [UNUserNotificationCenter currentNotificationCenter].delegate = self;
}

- (void)handleRemoteNotification:(id)notification
{
    if (!self.remoteNotificationHandler) return;
    
    NSDictionary *userInfo = nil;
    if ([notification isKindOfClass:[NSDictionary class]]) {
        userInfo = notification;
    } else {
        if ([notification isKindOfClass:[UNNotificationResponse class]]) {
            userInfo = ((UNNotificationResponse *)notification).notification.request.content.userInfo;
        } else if ([notification isKindOfClass:[UNNotification class]]) {
            userInfo = ((UNNotification *)notification).request.content.userInfo;
        }
    }
    
    self.remoteNotificationHandler(userInfo, notification);
}

- (void)handleLocalNotification:(id)notification
{
    if (!self.localNotificationHandler) return;
    
    NSDictionary *userInfo = nil;
    if ([notification isKindOfClass:[NSDictionary class]]) {
        userInfo = notification;
    } else {
        if ([notification isKindOfClass:[UNNotificationResponse class]]) {
            userInfo = ((UNNotificationResponse *)notification).notification.request.content.userInfo;
        } else if ([notification isKindOfClass:[UNNotification class]]) {
            userInfo = ((UNNotification *)notification).request.content.userInfo;
        }
    }
    
    self.localNotificationHandler(userInfo, notification);
}

#pragma mark - UNUserNotificationCenterDelegate

// 前台收到推送
- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler
{
    // 远程推送
    if ([notification.request.trigger isKindOfClass:[UNPushNotificationTrigger class]]) {
        completionHandler(UNNotificationPresentationOptionBadge | UNNotificationPresentationOptionSound | UNNotificationPresentationOptionAlert);
    // 本地推送
    } else {
        completionHandler(UNNotificationPresentationOptionBadge | UNNotificationPresentationOptionSound | UNNotificationPresentationOptionAlert);
    }
}

// 后台收到推送
- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
{
    // 远程推送
    if ([response.notification.request.trigger isKindOfClass:[UNPushNotificationTrigger class]]) {
        [self handleRemoteNotification:response];
        completionHandler();
    // 本地推送
    } else {
        [self handleLocalNotification:response];
        completionHandler();
    }
}

#pragma mark - Local

- (void)registerLocalNotification:(NSString *)identifier title:(NSString *)title subtitle:(NSString *)subtitle body:(NSString *)body userInfo:(NSDictionary *)userInfo badge:(NSInteger)badge soundName:(NSString *)soundName timeInterval:(NSInteger)timeInterval repeats:(BOOL)repeats block:(void (NS_NOESCAPE ^)(UNMutableNotificationContent *))block
{
    UNMutableNotificationContent *notification = [[UNMutableNotificationContent alloc] init];
    if (title) notification.title = title;
    if (subtitle) notification.subtitle = subtitle;
    if (body) notification.body = body;
    if (userInfo) notification.userInfo = userInfo;
    notification.badge = badge > 0 ? [NSNumber numberWithInteger:badge] : nil;
    if (soundName) notification.sound = [@"default" isEqualToString:soundName] ? [UNNotificationSound defaultSound] : [UNNotificationSound soundNamed:soundName];
    if (block) block(notification);
    
    UNNotificationTrigger *trigger = timeInterval > 0 ? [UNTimeIntervalNotificationTrigger triggerWithTimeInterval:timeInterval repeats:repeats] : nil;
    UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:identifier content:notification trigger:trigger];
    [[UNUserNotificationCenter currentNotificationCenter] addNotificationRequest:request withCompletionHandler:nil];
}

- (void)removeLocalNotification:(NSArray<NSString *> *)identifiers
{
    [[UNUserNotificationCenter currentNotificationCenter] removePendingNotificationRequestsWithIdentifiers:identifiers];
    [[UNUserNotificationCenter currentNotificationCenter] removeDeliveredNotificationsWithIdentifiers:identifiers];
}

- (void)removeAllLocalNotifications
{
    [[UNUserNotificationCenter currentNotificationCenter] removeAllPendingNotificationRequests];
    [[UNUserNotificationCenter currentNotificationCenter] removeAllDeliveredNotifications];
}

@end

File: FWFramework/Classes/Compatible/Service/FWNotification.swift
//
//  FWNotification.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/27.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import UserNotifications
#if FWMacroSPM
import FWFramework
#endif

/// 通知管理器
@objcMembers public class FWNotificationManager: NSObject, UNUserNotificationCenterDelegate {
    /// 单例模式
    public static let sharedInstance = FWNotificationManager()
    
    // MARK: - Authorize
    
    /// 异步查询通知权限状态，当前线程回调
    public func authorizeStatus(_ completion: ((FWAuthorizeStatus) -> Void)?) {
        FWAuthorizeManager.manager(type: .notifications)?.authorizeStatus?(completion)
    }
    
    /// 执行通知权限授权，主线程回调
    public func requestAuthorize(_ completion: ((FWAuthorizeStatus) -> Void)?) {
        FWAuthorizeManager.manager(type: .notifications)?.authorize(completion)
    }
    
    // MARK: - Badge
    
    /// 清空图标通知计数
    public func clearNotificationBadges() {
        UIApplication.shared.applicationIconBadgeNumber = 0
    }
    
    // MARK: - Handler
    
    /// 设置远程推送处理句柄，参数为userInfo和原始通知对象
    public var remoteNotificationHandler: (([AnyHashable: Any]?, Any) -> Void)?
    
    /// 设置本地推送处理句柄，参数为userInfo和原始通知对象
    public var localNotificationHandler: (([AnyHashable: Any]?, Any) -> Void)?
    
    /// 注册通知处理器，iOS10+生效，iOS10以下详见UIApplicationDelegate
    public func registerNotificationHandler() {
        UNUserNotificationCenter.current().delegate = self
    }
    
    /// 处理远程推送通知，支持NSDictionary|UNNotification|UNNotificationResponse
    public func handleRemoteNotification(_ notification: Any) {
        guard remoteNotificationHandler != nil else { return }
        
        var userInfo: [AnyHashable: Any]?
        if let dictionary = notification as? [AnyHashable: Any] {
            userInfo = dictionary
        } else if let response = notification as? UNNotificationResponse {
            userInfo = response.notification.request.content.userInfo
        } else if let notify = notification as? UNNotification {
            userInfo = notify.request.content.userInfo
        }
        remoteNotificationHandler?(userInfo, notification)
    }
    
    /// 处理本地通知，支持NSDictionary|UNNotification|UNNotificationResponse
    public func handleLocalNotification(_ notification: Any) {
        guard localNotificationHandler != nil else { return }
        
        var userInfo: [AnyHashable: Any]?
        if let dictionary = notification as? [AnyHashable: Any] {
            userInfo = dictionary
        } else if let response = notification as? UNNotificationResponse {
            userInfo = response.notification.request.content.userInfo
        } else if let notify = notification as? UNNotification {
            userInfo = notify.request.content.userInfo
        }
        localNotificationHandler?(userInfo, notification)
    }
    
    // MARK: - Local
    
    /// 注册本地通知，badge为0时不改变，soundName为default时为默认声音，timeInterval为触发时间间隔(0为立即触发)，block为自定义内容句柄，iOS15+支持时效性通知，需entitlements配置开启
    public func registerLocalNotification(_ identifier: String, title: String?, subtitle: String?, body: String?, userInfo: [AnyHashable: Any]?, badge: Int, soundName: String?, timeInterval: TimeInterval, repeats: Bool, block: ((UNMutableNotificationContent) -> Void)?) {
        let notification = UNMutableNotificationContent()
        if let title = title { notification.title = title }
        if let subtitle = subtitle { notification.subtitle = subtitle }
        if let body = body { notification.body = body }
        if let userInfo = userInfo { notification.userInfo = userInfo }
        notification.badge = badge > 0 ? NSNumber(value: badge) : nil
        if let soundName = soundName { notification.sound = (soundName == "default") ? .default : UNNotificationSound(named: UNNotificationSoundName(soundName)) }
        block?(notification)
        
        let trigger = timeInterval > 0 ? UNTimeIntervalNotificationTrigger(timeInterval: timeInterval, repeats: repeats) : nil
        let request = UNNotificationRequest(identifier: identifier, content: notification, trigger: trigger)
        UNUserNotificationCenter.current().add(request, withCompletionHandler: nil)
    }
    
    /// 批量删除本地通知(未发出和已发出)
    public func removeLocalNotification(_ identifiers: [String]) {
        UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: identifiers)
        UNUserNotificationCenter.current().removeDeliveredNotifications(withIdentifiers: identifiers)
    }
    
    /// 删除所有本地通知(未发出和已发出)
    public func removeAllLocalNotifications() {
        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
        UNUserNotificationCenter.current().removeAllDeliveredNotifications()
    }
    
    // MARK: - UNUserNotificationCenterDelegate
    /// 前台收到推送
    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        // 远程推送
        if notification.request.trigger is UNPushNotificationTrigger {
            completionHandler([.badge, .sound, .alert])
        // 本地推送
        } else {
            completionHandler([.badge, .sound, .alert])
        }
    }
    
    /// 后台收到推送
    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        // 远程推送
        if response.notification.request.trigger is UNPushNotificationTrigger {
            handleRemoteNotification(response)
            completionHandler()
        // 本地推送
        } else {
            handleLocalNotification(response)
            completionHandler()
        }
    }
}
